src/main/scala/com/wavesplatform/network/Coordinator.scala

```diff
@@ -193,134 +145,87 @@ class Coordinator(protected override val networkControllerRef: ActorRef, blockch
-        items.exists { case BlockCheckpoint(h, sig) =>
-          h == estimatedHeight && (blockSignature != ByteStr(sig))
-        }
+  private def makeBlockchainCompliantWith(checkpoint: Checkpoint): Unit = {
+    val existingItems = checkpoint.items.filter {
+      checkpoint => history.blockAt(checkpoint.height).isDefined
     }
 
-  private def validateWithRespectToCheckpoint(candidate: Block, estimatedHeight: Int): Either[ValidationError, Unit] = {
-    if (isValidWithRespectToCheckpoint(candidate, estimatedHeight))
-      Right(())
-    else
-      Left(GenericError(s"Block ${str(candidate)} [h = $estimatedHeight] is not valid with respect to checkpoint"))
-  }
-
-  def isBlockValid(b: Block): Either[ValidationError, Unit] = {
-    if (history.contains(b)) Right(())
-    else {
-      def historyContainsParent = history.contains(b.reference)
-
-      def consensusDataIsValid = blockConsensusValidation(
-        history,
-        stateReader,
-        settings.blockchainSettings,
-        time)(b)
+    val fork = existingItems.takeWhile {
+      case BlockCheckpoint(h, sig) =>
+        val block = history.blockAt(h).get
+        block.signerData.signature != ByteStr(sig)
+    }
 
-      if (!historyContainsParent) Left(GenericError(s"Invalid block ${b.encodedId}: no parent block in history"))
-      else if (!Signed.validateSignatures(b).isRight) Left(GenericError(s"Invalid block ${b.encodedId}: signature is not valid"))
-      else if (!consensusDataIsValid) Left(GenericError(s"Invalid block ${b.encodedId}: consensus data is not valid"))
-      else Right(())
+    if (fork.nonEmpty) {
+      val genesisBlockHeight = 1
+      val hh = existingItems.map(_.height) :+ genesisBlockHeight
+      history.blockAt(hh(fork.size)).foreach {
+        lastValidBlock =>
+          log.warn(s"Fork detected (length = ${fork.size}), rollback to last valid block id [${lastValidBlock.encodedId}]")
+          blockchainUpdater.removeAfter(lastValidBlock.uniqueId)
+      }
     }
   }
 
-
-  private def processNewBlock(block: Block): Either[ValidationError, Unit] = for {
-    _ <- validateWithRespectToCheckpoint(block, history.height() + 1)
-    _ <- isBlockValid(block)
-    _ <- blockchainUpdater.processBlock(block)
-  } yield {
-    block.transactionData.foreach(utxStorage.remove)
-    UnconfirmedTransactionsStorage.clearIncorrectTransactions(settings.blockchainSettings.functionalitySettings,
-      stateReader, utxStorage, time)
+  private def checkExpiry(ctx: ChannelHandlerContext): Unit = {
+    if ((time.correctedTime() - history.lastBlock.timestamp).millis > 1.hour) {
+      ctx.writeAndFlush(BlockchainExpired)
+    } else {
+      ctx.writeAndFlush(BlockchainUpdated)
+    }
   }
 
-
-  private def str(block: Block) = {
-    if (log.logger.isTraceEnabled) block.json
-    else block.uniqueId + ", parent " + block.reference
+  override def channelRead(ctx: ChannelHandlerContext, msg: AnyRef) = msg match {
```

Comment:
`Coordinator` is about business logic, `channelRead` is about implementation detail, also non-type-sage, they should be separated and `Coordinator`  needs to be referencened in channel reader.